scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +172,393 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
         i += 1
 
     return np.reshape(x, size1d)
+
+
+class FastNumericalInverse():
+    r"""
+    Object representing a Fast Numerical Inverse of a probability distribution.
+
+    The initializer of `FastNumericalInverse` accepts `dist`, a frozen instance
+    of a continuous distribution, and provides an object with methods
+    that approximate `dist.ppf` and `dist.rvs`. For most distributions,
+    these methods are faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
```

Comment:
Because the method also wants the PDF (slope of the spline), and I use the PPF and ISF functions (at one point each) to get the lower and upper ends of the domain over which to create the mesh. An `rv_frozen` packages all this nicely, and if the user has only the PDF, they can make use of the `rv_continuous` infrastructure to generate the rest rather than providing them all.

Reply:
Thanks. Two comments/questions:

- do we need isf and ppf if the domain is bounded? I think one can then directly use the endpoints unless the pdf is zero. this might save an evaluation of ppf and isf.
- If I want to use the inversion method to speed up a distribution, would that work well with the current interface? Say you define `class mydistr(rv_continuous)` and in the init step, you want to run the inversion and save it (e.g. self._fni = fastNumericalInverse(...)) and the use it in the rvs method: `def _rvs(self, ...): return self._fni.rvs(...)`, same for ppf. At the moment, wouldn't I need to create another frozen rv to call `fastNumericalInverse`. I haven't thought about it in detail but that would be a common use case.

Reply:
1) We don't. I thought about that, but only after others started reviewing, so I thought I should add it in a follow-up. Easy enough to do, so if @tupui doesn't mind the change, I could do it later this week.

2) If we would want to use it in the RVS method of a non-frozen distribution (which would only really help `genexpon` and the proposed `studentized_range`, at this point), then there is no problem. I just did a qucik and dirty check with `genexpon` like:
```python3
    def rvs(self, a, b, c, size=1):
        return stats.FastNumericalInverse(self(a, b, c)).rvs(size)
```
Results:
```
%timeit stats.genexpon.rvs(a, b, c, size = 1000)  # before
3.38 s ± 260 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit stats.genexpon.rvs(a, b, c, size = 1000)  # after
20.3 ms ± 628 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
I think it's more likely we would want to use this only in the `rvs` method of frozen distributions, and creating the FNI in the `__init__` method like:
```python3
        self.fni = stats.FastNumericalInverse(self)
        self.rvs = lambda size: self.fni.rvs(size)  # this override the original `rvs` method
        self.ppf = lambda q: self.fni.ppf(q)
```
Then:
```python3
a, b, c = (9.13, 16.2, 3.28)
from scipy import stats
import numpy as np
dist = stats.genexpon(a, b, c)
%timeit dist.ppf(np.linspace(0, 0.99, 1000))  # 166 µs ± 4.36 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
I don't mind the changes, these seem sensible and reasonable. I would just make sure the timeline is ok for you as the branching is soon (although personally I would even wait for that to be done. Having this would really help me to advertise the QMC module to the QMC people in august.). The performance gains are quite impressive! 👏

Reply:
> Perhaps we should always allow the user to avoid the tails?

That would be great. Maybe, add a parameter `domain` that accepts the user-defined domain of the distribution?

Reply:
The current interface is similar to the one in RUnuran (R package):

```
## Compute quantiles of user-defined distribution using method 'HINV'
## (using 'advanced' interface)
cdf <- function (x) { 1.-exp(-x) }
pdf <- function (x) { exp(-x) }
dist <- new("unuran.cont", cdf=cdf, pdf=pdf, lb=0, ub=Inf)
gen <- unuran.new(dist, "hinv; u_resolution=1.e-12")
uq(gen,seq(0,1,0.05)) # compute quantiles
```

I think the interface is convenient if I have a custom distribution in mind and I want to try the inversion method. All I need to do is define a rv_continuous object with `_pdf` and `_cdf`.

If we want to speed up the _rvs method of a distribution, I agree that this should be done for frozen distributions, otherwise one has to compute the inversion every time if rvs is called. @mdhaber Could you paste the full code that you used to make that work for frozen genexpon above please?

Reply:
I don't have any more code, really. I could make a fork ( I got rid of the changes after testing then), but I think it will just work if you paste the code above at the end of the `rv_frozen` init method. 

More code (e.g. `if` statement that looks at the distribution being frozen) would be required to do this only for certain distributions, but i didn't write it just to test the concept. That's not a problem with this function accepting an `rv_continuous` - it's just that there's currently no way to override a particular distribution's frozen behavior.

Reply:
@chrisb83 I created a branch [`fni_rv_frozen`](https://github.com/mdhaber/scipy/tree/fni_rv_frozen) for you to test in which the `rvs` and `ppf` method of all frozen distributions is replaced with those of the numerical inverse. As-is, think this will break discrete distributions, but it's easy to add conditions that would only make the change to certain continuous distributions.

Reply:
Thanks and sorry for asking lots of questions about this. If I understand correctly, this means that if we want to use the method for a distribution already implemented in SciPy, we would need to modify the `rv_frozen` class and add something like `if name of distribution in inversion_list: self.fni = ...`. 

If a user want to create a custom distribution, there is no easy way to create a frozen distribution for it that uses the inversion method since it would require to change the rv_frozen class (or create another class that inherits from it that the user can then modify). Or do I miss something here?

Tirth also pointed out that using the pdf/cdf methods of the distribution object can create quite a bit of overhead when he tested it, so he opted to use a function as a parameter.

I just looked at all the univarite distribution methods in UNU.RAN: most of them just require one function as an input (usually the pdf), at most there are three functions (HINV, PINV, depending on the settings. e.g. for HINV, one needs the derivative of the PDF if quintic Hermite polynomials are used, PINV accepts CDF but only as an experimental feature. the derivative of the PDF is anyways not part of rv_continuous). If two functions are required, one is often the derivative of the pdf that is not in rv_continuous.

So my feeling is that we are overall better of if we have a consistent interface with parameters like cdf, pdf, dpdf, logpdf etc.

One can than still write `NumericalInversionHermite(pdf=dist.pdf, lb=dist.ppdf(...), ub=dist.isf(...))` (lb/ub = lower/upper bound of the domain). 

Please let me know what you think. FYI @mckib2

Reply:
That's a good point. This should have an interface (but possibly not the only interface) that can be used inside of `_rvs()` and `_ppf()` method of distributions. We'd like to use this in some current distributions that are quite slow without having the user have to invoke `NumericalInversionHermite` themselves.

That said, it would also be handy to also have an interface that just takes a frozen distribution and pulls the methods it needs from it. `@classmethod`s are nice for this: `NumericalInversionHermite.fromdist(some_frozen_dist)` would work as an alternate constructor.

Reply:
> If a user want to create a custom distribution, there is no easy way to create a frozen distribution for it that uses the inversion method since it would require to change the rv_frozen class (or create another class that inherits from it that the user can then modify). Or do I miss something here?

I think you're right, but I think this is totally independent of whether an `rv_frozen` is required as a parameter. (I don't see how any other interface to `FastNumericalInverse` would make it easier for users to create their own custom distributions with corresponding `rv_frozen` that has fast-numerical-inverse-based `rvs` and `ppf` methods.)

> Tirth also pointed out that using the pdf/cdf methods of the distribution object can create quite a bit of overhead when he tested it, so he opted to use a function as a parameter.

It definitely does. I mentioned that we can open this up in the future so that the user doesn't need to pass in a frozen distribution - just an object that has a CDF method at a minimum, and if desired we could wrap that to get the PDF method needed for the slope of the spline. We could also wrap the CDF to get PPF/ISF methods, or if we allow the user to pass a domain, then we wouldn't need PPF/SF methods. 

The only reason the object _has_ to be an instance of RV frozen right now is because it was an easy way of doing input validation and documentation. We can open that up right now if someone is willing to write the extra documentation.

> So my feeling is that we are overall better of if we have a consistent interface with parameters like cdf, pdf, dpdf, logpdf etc.

Isn't it just as convenient to instead accept a single object with these methods rather than four separate parameters, one for each function? Internally, you just unpack the separate methods from object using `getattr` (e.g. `pdf = getattr(dist, "pdf")`). If a user is going to create `cdf`, `pdf`, `dpdf` (which, BTW, I think we should add to `rv_continuous` because it could be used in fitting), and `logpdf`, it is reasonable that they should have some sort of object to hold all that together. (`rv_continuous` was created because it was convenient to group functions related to a given continuous distribution together, right?)

If we really want to accept separate functions as arguments, they could be added in a backwards compatible way in the future if we set a default value of the `dist` argument (`None`) and add keyword arguments for all these separate functions.

@rkern 
> This should have an interface (but possibly not the only interface) that can be used inside of _rvs() and _ppf() method of distributions. We'd like to use this in some current distributions that are quite slow without having the user have to invoke NumericalInversionHermite themselves

I showed above how the current interface can already do that. The current interface also addresses your point "That said, it would also be handy to also have an interface that just takes a frozen distribution..."

TLDR:
It seems there is a lot of confusion caused by the requirement that the `dist` parameter has to be an instance of `rv_frozen`. First, this isn't a real requirement of the implementation; it just made documentation and input validation easier. Currently, `FastNumericalInverse` just needs an object with `pdf`, `cdf`, `ppf`, and `isf` methods. The requirement can be relaxed in the future to accept _any_ sort of object with these methods - or a subset of these methods, if not all are available. If it is really desirable, we could give the first initializer argument `dist` a default value and add keyword arguments that accept separate functions `cdf`, `pdf`, etc... The current interface can be used as-is to override the `rvs`/`ppf` methods of both `rv_continuous` or `rv_frozen` classes, and the fact that `dist` has to be an `rv_frozen` doesn't seem to get in the way. (I'm not seeing circular reference sorts of problems, if that's the concern.) All of the comments about allowing the user to specify a custom domain or only requiring the user to provide a `pdf` are great enhancement ideas. I've had them in mind, too.  These enhancements can be added in a backwards compatible way, and I don't think we'll end up with a clumsier interface if we start with this class as-is and add these features later.

Reply:
> I showed above how the current interface can already do that.

Do you mean the [`fni_rv_frozen`](https://github.com/mdhaber/scipy/tree/fni_rv_frozen) branch? That's not the right place to implement that functionality. People do use the unfrozen distribution objects.

Reply:
In [this comment](https://github.com/scipy/scipy/pull/13319#discussion_r634500334) I showed how it could be done for un-frozen distribution objects, too:
```
    def rvs(self, a, b, c, size=1):
        return stats.FastNumericalInverse(self(a, b, c)).rvs(size)
```

Reply:
That is _a_ way to get the computation done, but I don't think it's a good design choice. Architectures that respect a well-ordered layering are generally easier to work with.

Reply:
"In this comment I showed how it _could_ be done..." I didn't claim it was the best way. 

If you prefer, we could directly call `_fast_numerical_inverse` and provide it _any object that has the required methods_. It does not have to be a frozen instance of `self` if that's a problem. `_fast_numerical_inverse` is private, so we can we can change its interface however we please in the future to better support whatever use we have in mind.

I don't think we need to these decisions in this PR. The primary intended purposes of this class, as written, was 1) to allow users to create faster PPF and RVS methods for existing distributions and 2) allow users to draw quasi-random variates from existing distributions. Is this an interface we want to expose in this release or not?